[TG-6261] Add option to make numerical input values integral

majakusber opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

This introduces new option `--java-assume-inputs-integral` that forces all generated numerical values to be integral, not fractional. Namely, it forces all `double`s and `float`s to have `int` value. 

The implementation adds an assume after nondet-ing the the original type. Alternative solution would be for `double`s and `float`s  to directly nondet the expression to an `int`. I believe the solution with an assume is more clean (does not change the flow of nondet-ing in object factory, only adds assume) and easier to understand. Let me know in comments if you have a different opinion.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj commented on jbmc/src/java_bytecode/java_bytecode_language.h:
```diff
@@ -62,6 +63,10 @@ Author: Daniel Kroening, kroening@kroening.com
+  " --java-assume-inputs-integral\n" \
+  "                              only initialize numerical primitive-typed parameter\n" /* NOLINT(*) */ \
+  "                              to an entry point that is integral, not fractional;\n" /* NOLINT(*) */ \
+  "                              results in float and double having int value\n" /* NOLINT(*) */ \
```
:no_entry_sign: Also needs documentation here: https://github.com/diffblue/test-gen/blob/develop/doc/cl-options.md
with a short example, ideally.
The limitation (arrays) should also be stated there.

majakusber replied:
Definitely, that will be part of the test-gen PR.

majakusber replied:
I added the limitation on arrays

allredj replied:
Hmmm... of course, that PR is indeed on cbmc... :man_facepalming:

allredj commented on jbmc/src/java_bytecode/java_bytecode_language.h:
```diff
@@ -62,6 +63,10 @@ Author: Daniel Kroening, kroening@kroening.com
   "                              and upper bound U must be integers\n" /* NOLINT(*) */ \
+  " --java-assume-inputs-integral\n" \
+  "                              only initialize numerical primitive-typed parameter\n" /* NOLINT(*) */ \
+  "                              to an entry point that is integral, not fractional;\n" /* NOLINT(*) */ \
```
:pick: I would avoid using `fractional` here, as it's not the best way of describing floats (even if the can be seen as fractional from a mathematical point of view).
I would also avoid the notion of `entry-point`, as it can be confused with the entry-point of the analysis.

:no_entry_sign: Also, `parameter` is not quite sufficient, as this option also affects class fields.
I would simply write:
```
Force float and double inputs to have integer value.
```

majakusber replied:
Done

allredj commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -901,6 +901,45 @@ void java_object_factoryt::gen_nondet_struct_init(
+    java_int_type(),
+    "assume_integral_tmp",
+    location,
+    "assume_integral_tmp",
```
Why `_tmp`?

majakusber replied:
Just to indicate that the symbol is very local, it's created just for the purpose of the assume and is destroyed right after (the `dead` statement appears right after the assume in the goto program).

allredj commented on jbmc/src/java_bytecode/java_object_factory_parameters.h:
```diff
@@ -25,6 +25,9 @@ struct java_object_factory_parameterst final : public object_factory_parameterst
   /// Force numerical primitive inputs to fall within the interval
   integer_intervalt assume_inputs_interval;
 
+  /// Force numerical primitive inputs to be integral (not fractional)
```
remove `(not fractional)`

majakusber replied:
Done

allredj reviewed (changes requested):
Implementation looks great to me. Some documentation requests (as always). :-)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 559e1bf865df94083f089f9155b2e2ed5480d44e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117011328

allredj commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1011,8 +1050,7 @@ void java_object_factoryt::gen_nondet_init(
     assignments.add(assign);
-    // add assumes to obey numerical limits (given via
-    // java-assume-inputs-interval)
+    // add assumes to obey numerical restrictions
```
:pick: Some documentation in the function doc would be useful:
```
Extra constraints are added to initialized objects, as specified in`object_factory_parameters`.
```

majakusber replied:
What is the link supposed to point to?

allredj replied:
It's the field you get the options from.

majakusber replied:
Added

allredj commented on jbmc/src/java_bytecode/java_object_factory_parameters.h:
```diff
@@ -25,6 +25,9 @@ struct java_object_factory_parameterst final : public object_factory_parameterst
   /// Force numerical primitive inputs to fall within the interval
   integer_intervalt assume_inputs_interval;
 
+  /// Force double and float inputs to be integral
```
:pick: Add `This does not apply to array elements`.

majakusber replied:
I wouldn't propagate the restrictions everywhere, it's in the option documentation.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e669c37ff57b272bfe0d0773a220c3187ea9ddb5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117128656
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

majakusber commented:
Test-gen PR: https://github.com/diffblue/test-gen/pull/3605

thk123 commented on jbmc/regression/jbmc/assume-inputs-integral/My.java:
```diff
@@ -0,0 +1,84 @@
+      assert false;
+    } else if (doubleArg == 1.08) {
+      assert false;
+    } else if (doubleArg != floatArg) {
```
You should test that it doesn't allow Float.NaN, Float.NEGATIVE_INFINITY Float.POSITIVE_INFINITY

majakusber replied:
Added :+1:

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 8fd87c09a177e2e2ffb691faf4c797df1690c3fc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117186603
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c8a8aef97675e4f9024d2d2570a699c0a2018b10).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117190377
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4d0d62db4428ecfe2b98a42b813dbd62ac124915).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117192152

thk123 commented on jbmc/regression/jbmc/assume-inputs-integral/class_arr_assume.desc:
```diff
@@ -0,0 +1,29 @@
+Check that --java-assume-inputs-integral restricts double and float arrays to
+integral values.
+
+This does not work yet, see TG-8291.
```
:question: Is it acceptable to put jira references in cbmc? @peterschrammel ?

majakusber replied:
Good point, there are only a couple of ticket numbers in `regression/jbmc` folder so I'll remove these (they will be in test-gen tests).

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -894,11 +894,53 @@ void java_object_factoryt::gen_nondet_struct_init(
+    symbol_table);
+  const auto &aux_expr = aux_symbol.symbol_expr();
+  assignments.add(code_declt(aux_expr), location);
+  exprt aux_rhs = side_effect_expr_nondett(java_int_type(), location);
```
:bulb: Might want to call this `nondet` rather than `aux` as a bit more informative

majakusber replied:
Done

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -894,11 +894,53 @@ void java_object_factoryt::gen_nondet_struct_init(
+  const exprt &expr,
+  const typet &type,
+  const source_locationt &location,
+  symbol_table_baset &symbol_table)
```
:bulb: You might like to take a `allocate_local_symbolt` (in `nondet.h`) rather than a mutable symbol table to make the code more easily reusable in smart harness.

majakusber replied:
Cool, thank you! Done

majakusber replied:
I left this change in a separate commit for easier reviewing

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -894,11 +894,53 @@ void java_object_factoryt::gen_nondet_struct_init(
+  test_assign.add_source_location() = location;
+  assignments.add(test_assign);
+  assignments.add(code_assumet(
+    binary_relation_exprt(typecast_exprt(aux_expr, type), ID_equal, expr)));
```
:pick: You can use `equal_exprt` to represent an equal binary relation.

majakusber replied:
Done

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c87a577acd22b0b4b5b25eeec266bdae4db0b655).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117295592
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 62afcf5f1f3a9ba992c0bdec2aa49aab69135bc6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117299129

Commit:
Use allocate_local_symbolt instead of passing symbol_table

It's cleaner than passing a mutable symbol table and it makes the code more
easily reusable in smart harness.

Commit:
Add --java-assume-inputs-integral option

Commit:
Add regression tests for --java-assume-inputs-integral

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6f50f5791198b8f96cd04ed252fc002c72744846).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117485004

allredj commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -899,26 +899,22 @@ void java_object_factoryt::gen_nondet_struct_init(
 /// \param location Source location associated with the expression
-/// \param [out] symbol_table Symbol table, a new auxiliary symbol will be added
+/// \param allocate_local_symbol A function to generate a new local symbol
+///   and add it to the symbol table
```
:pick: Maybe say "why" in commit message

majakusber replied:
Done

allredj reviewed (approved):
Still good.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a69208c7944457f7fe6c75462b91bf9d2d06caed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117529733
